### DISCUSSION TITLE
Adapting the plugin for usage in Firefox-based browsers

### DISCUSSION THREAD

**Discussion by ghost** [keybraker/reSkroutzed#5] [Ideas]
Greetings, all.

I forked the project yesterday, and I plan to port the extension to Firefox by making some small modifications. With the – relatively miniscule – amounts of testing I've done by now, the extension seems to mostly work already, albeit with a few (primarily visual) bugs and API incompatibilities that should be simple-enough to fix.

However, that raises a few questions that I'd like to ask, assuming that I *do* make the extension fully work on Firefox before the current primary contributors do:

* For security reasons, the only Firefox versions that support unsigned (by Mozilla) visual extensions are Firefox Beta, Firefox Nightly, and Firefox Developer Edition. As such, if the average user wants to install a visual extension, the extension will need to be submitted to the Mozilla store and be manually approved by someone from Mozilla. 
  - Who shall be the submitter of the app to the firefox marketplace? 
  - Who should be mentioned as the author in the "browser-specific" section of the firefox Manifest file?

* What shall be done with the updated (for Firefox) source code? Should it be committed to a special branch of the original repo, a special folder in the main branch of the original repo, or should it stay in my fork?

Seeing as the extension is GNU GPLv3 licenced, standard GPLv3 terms apply, but I would still prefer to ask for the opinion of the original maintainers! :)

**Comment by keybraker** [comment]
Hello there and thank you for your interest in the repo,

My main question is, whether the changes made it incompatible with chrome one? or better yet, if a properly working chrome and Firefox extension, are compatible between them. 

I wish to keep a single source of truth for the source code, instead of mainting multiple repos/forks.

The release will be made by this account on Firefox as well.

**Comment by ghost** [comment]
Hello!

JS-wise, the extension mostly seems to work; the Firefox-related bugs I've found by now are mostly visual (i.e. everything seems to be left-aligned, some elements get pushed outside the screen, a part of the screen remains empty of content…). I didn't get to fix this yet, but it shouldn't be that complicated. It *might* be possible for this to be patched in a way that doesn't break cross-compatibility with Chromium, but we shall see when I get to that.

The [Firefox cross-compatibility tool](https://www.extensiontest.com/) currently reports the following:

```js
{
  "compat": [],
  "errors": [
    {
      "message": "\"/manifest_version\" must be <= 2",
      "description": "Your JSON file could not be parsed.",
      "locations": [
        {}
      ]
    }
  ],
  "warnings": [
    {
      "message": "Unsafe assignment to innerHTML",
      "description": "Due to both security and performance concerns, this may not be set using dynamic values which have not been adequately sanitized. This can lead to security issues or fairly serious performance degradation.",
      "locations": [
        {
          "file": "manipulators/addBlockedIndication.js",
          "line": 17
        },
        {
          "file": "manipulators/productFlagger.js",
          "line": 30
        },
        {
          "file": "manipulators/separatePromoListFlagger.js",
          "line": 24
        },
        {
          "file": "manipulators/shelfFlagger.js",
          "line": 22
        }
      ]
    }
  ]
}
```

However, the (single) error is misleading, as Firefox has been supporting ManifestV3 since v109, which was released in January 2023.

After I *do* patch these visual bugs, I'll just have to do a lot of testing on Firefox and a chromium-based browser, to ensure that they both behave identically. 

[You can currently see a sneak-peek here](https://drive.google.com/drive/folders/1r4bAy9yFV03kA9ARvDzc6teZCIQcYonu?usp=share_link), comparing between Firefox and Vivaldi (Chromium-based). Interestingly, a certain sponsored banner actually doesn't get flagged on Vivaldi but does so on Firefox (!) I'll definitely investigate this as well.

Regardless, the three most important things to note are:
* Firefox was treating the extension as "corrupted" until I added these few lines on the manifest file:
```json
  "browser_specific_settings": {
    "gecko": {
      "id": "addon@example.com",
      "strict_min_version": "109.0"
    }
```
* Chromium uses the "chrome.\*" namespace, whereas vanilla Firefox treats "chrome.\*" as an alias to the "browser.\*" namespace. This, however, is not an official standard and may not be supported on forks of Firefox
* ~~The "tabs" API behaves slightly differently, but I will investigate whether this is an issue.~~

In any case, I shall keep you updated with any new info. I assume I'll be done with the CSS-related patches by next weekend & begin properly testing.

**Comment by ghost** [comment]
Update: At its current state in my fork, the extension appears to work on both browsers, with minimal edits. In specific, JS-wise everything is cross-compatible, and the primary issue was the following line on the CSS file `popup.css`:

```css
.content {
  margin: 0;
}
```

Which I have now replaced with:

```css
.content {
  margin: auto;
}
```

The `manifest.json` file will most likely require us to keep the following lines:

```js
  "browser_specific_settings": {
    "gecko": {
      "id": "addon.todo@example.com",
      "strict_min_version": "109.0"
    }
  }
```

As [all Manifest V3 extensions need an add-on ID in their manifest.json when submitted to AMO.](https://extensionworkshop.com/documentation/develop/extensions-and-the-add-on-id/#when-do-you-need-an-add-on-id).

Currently, the plans are for me to:
* Thoroughly test for bugs on Firefox and/or incompatibilities 
* Investigate *why* the sponsored shelf doesn't appear to get flagged on Vivaldi (Chromium)

Then we can see whether it's going to be merged on the main branch or another branch :)

**Reply by keybraker** [reply]
Great news and thank you!

Will be merging some updates soon which should not affect your changes apart from version.
After you rebase your code I will take a look into it so we can finalise it and merge it and then release for Firefox as well.

Also from what I have briefly seen, in your updates `.content {  margin: auto; }` is something I try to avoid, if  `.content {  margin: 0px; }` would work as desired, I think it would be a better way forward.

**Reply by ghost** [reply]
Thankfully, there barely was anything that required editing! Especially JS-wise :) 
If I *do* edit the JS for compatibility, I'll be because of some different bug fix 😁

We should be able to find a workaround - the "margin bug" *might* be a result of the different treatment of paths on chromium vs firefox.

I'll let you know if there's any update, as currently, my few edits apparently only had to be on CSS and JSON files!

**Reply by ghost** [reply]
That is, I *had* edited the JS part originally, but it wasn't needed for this purpose, so I kept only the relevant commits now :)

**Reply by ghost** [reply]
It appears that the cause of the visual bug is that Firefox tries to apply `popup.css` to the Skroutz website because it's included in the manifest.json file here:

```json
  "content_scripts": [
    {
      "matches": ["https://www.skroutz.gr/*"],
      "css": [
        "css/flaggerToogle.css",
        "css/popup.css",
        "css/productFlags.css",
        "css/shelfFlags.css",
        "css/style.css",
        "css/videoFlags.css"
      ],
```

`popup.js` isn't included there, so perhaps this was accidental? Removing it solves the visual problem. Renaming the html class from `content` to `popup-content` appears to work as well, but that feels more like a hack than a proper solution IMO.

In other news, I may or may not have found a Firefox-related JS bug that I'll need to fix after your most recent commit. I'll properly report or fix it after I properly test that.

**Reply by keybraker** [reply]
Hey @bandisast, I removed `popup.css` from the latest version.
I am finishing up, with the changes for 1.4 and will shortly proceed with releasing it to the chrome store. Have you finished with all the modifications, to make it compatible with Firefox? If so please make a pr so we can have a double release on the Firefox store as well.

**Reply by ghost** [reply]
Hey!

I haven't yet tested whether any of the most recent changes broke firefox compatibility, but I will do so at the soonest convenience. If no issue is found, I suspect that the only leftover change will be the `browser_specific_settings` section, seeing as `popup.css` was already removed & I haven't discovered any firefox-specific bugs due to the extension being relatively simple and "lightweight" :)

I'll test the newest commits and I'll let you know if I find anything else that might delay a PR, otherwise only expect a minor edit in the `manifest.json` file.

**Reply by keybraker** [reply]
Great to hear and thanks!
Will take a look on how to deploy on Firefox later today, in order to schedule a release submission for both on Friday the 28/04/23.

**Comment by keybraker** [comment]
Just to inform you guys, I have switched to a typescript and added webpack as the build process. It is documented in the `README.md` file so you can take a look on how to build it.

There are now two different build processes for Chrome and Firefox, so they are developed in parallel but also independently. 
One last note is that I have released the extension to the Firefox Add-ons store on version 1.4, were the review is pending.

**Reply by ghost** [reply]
That's great news, thanks for the update :)

Paging @gtopsis so that they see this as well

**Reply by gtopsis** [reply]
Awesome news. Very nice initiative to support FF ❤️  friends @bandisast.
Very professional work @keybraker with the ts and webpack parts @keybraker. 
[Vite](https://vitejs.dev/) may be a good replacement in the far future :P
